### DISCUSSION TITLE
Search / replace selected text only?

### DISCUSSION THREAD

**Discussion by zambetti** [zed-industries/zed#8608] [Feature Requests]
I suspect Zed does not support search/replace in 'selected text only', since when you click the magnifying glass while text is selected, it assumes that you are searching for that text.

This assumption might speed up some scenarios, however this makes it impossible to support a very useful variant of search/replace, one that is constrained to the selected text. Many editors offer this, notably the venerable BBEdit.

So it seems as it is now, Zed can only perform search/replace on an entire document, and there is no way to constrain search/replace to only the selected text. Can someone confirm this?

**Comment by mrnugget** [comment]
Yup, that's the current state.

**Reply by zambetti** [reply]
Thank you for the quick answer, it is good to know that this is not supported.

A workaround in the meantime is to cut/paste my selection into a new buffer, search/replace all, then cut/paste it back where it was in the original buffer. Not ideal, but it works in a pinch.

Do you feel that selection-constrained search/replace is worthwhile proposing as a feature/improvement?
If so, would you suggest that I create a 'Feature Request' issue?

**Reply by mrnugget** [reply]
Yeah, we can create a ticket for that.

FWIW I do think that multi-cursors go a long way here already. It requires kind of a mindset shift, but I'm starting to see it more and more.

**Reply by zambetti** [reply]
Thank you for this suggestion, I wasn't aware that multi-cursor could be used in this way via cmd-d. This will work well for simpler scenarios where I would have used selection-constrained search/replace in the past.

Another common situation where I use selection-constrained search/replace, is for re-ordering parameters (or other separated text) via regex match groups and numbered backreferences, but this does not seem to be supported by Zed's search/replace when Regex is enabled.

Support for numbered backreferences in Regex replacement is a different topic than selection-constrained search/replace, but I believe that there would be no other way to do this kind of edit in a non-document-wide fashion otherwise.

For example...
![Screenshot 2024-02-29 at 15 54 39](https://github.com/zed-industries/zed/assets/98515/d0f467cf-cc65-45a8-8aaa-bbcf746ef96c)

Results in...
![Screenshot 2024-02-29 at 15 54 49](https://github.com/zed-industries/zed/assets/98515/c267623d-37da-4be6-b305-14d2f325b2ce)

Instead of...
![Screenshot 2024-02-29 at 16 00 11](https://github.com/zed-industries/zed/assets/98515/ece1a28f-4c29-41cd-8c64-3d73f88c6231)

*** Update ***

I see there are already issues [#8315](https://github.com/zed-industries/zed/issues/8315) and [#7999](https://github.com/zed-industries/zed/issues/7999) on the topic of support for numbered backreferences in Regex replacement.

I've opened a feature request issue [#8617](https://github.com/zed-industries/zed/issues/8617)

**Comment by dream-dasher** [comment]
This is something in [Helix](https://helix-editor.com) does so well it sometimes think about going back just for it.

Especially with exploratory code it's very common that I just want to highlight a section, type some regex pattern (selecting that code), and then operate on all the highlighted code.

It's super common when doing exploration with debug elements, for example.
"Multi-cursor" is only useful if you can get the cursor where you want it.  If text isn't nicely aligned it's time consuming.
Much nicer to just highlight and then enter a pattern to determine where cursors are.

**Reply by Vehmloewff** [reply]
Yes, this was a very valuable feature to me in Helix. Super useful!

**Comment by failable** [comment]
I miss that in Emacs too. And Emacs has the ability to [narrow](https://www.gnu.org/software/emacs/manual/html_node/emacs/Narrowing.html) down to region and perform any text manipulation on that region.

**Comment by ddbrierton** [comment]
FWIW, something close to this is already available in Vim-Mode. Not automatically by selection, but you can limit a search and replace to a subset of the document, lines 39-45 like this: `:39,45s/foo/bar/`. So Zed has 90% of this functionality already, they just need to make it more generally available.

**Comment by ddbrierton** [comment]
In fact, I just realised that, further to my previous comment, Vim-Mode in Zed does support search and replace only within selected text. If you select some text in Vim-Mode and then press `:` the command palette is auto-populated with Vim's marks for the beginning and end of a selection:

```
`<,`>
```

all you then need to do is complete that with, for example

```
`<,`>s/foo/bar/
```

and Zed will replace all occurrences of "foo" with "bar" within the selection.

So Zed has this capability already. I guess they just haven't provided a UI for it in non-Vim-Mode yet.

**Reply by SomeoneToIgnore** [reply]
> haven't provided a UI for it in non-Vim-Mode yet

What's this?

<img width="416" alt="image" src="https://github.com/user-attachments/assets/f84758e2-9577-4754-9f7a-3b7f537defbd" />

**Reply by ddbrierton** [reply]
Oh! Ha ha, you're absolutely right. (I hadn't noticed it because I use Vim-Mode.)

But then that should be the answer to the entire thread.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
